### DISCUSSION TITLE
In the array / variable search, how most effective ?

### DISCUSSION THREAD

**Discussion by StefanSa** [vectordotdev/vector#10791] [Q&A] [answered]
I have a syslog device that does not send the hostname, but the serial number. Currently i create the hostname in number of devices over several "if" queries.
I would like to have this simpler if possible, but do not want to solve this via the enrichment option.
I imagine for example a variable / array like this:
.hosts =["SN01": "host01", "SN02": "host02"]

My question now how to search in this variable / array for the SN and then assign it to the matching host ?

**Comment by spencergilbert** [comment, marked as answer]
Slightly different from your existing example, but with this example event:

```coffee
$ .
{ "message": "foobar", "serialnumber": "SN01" }
```

And a variable in VRL set to an object of serialnumber => hostname k/v's

```coffee
$ hosts
{ "SN01": "host01", "SN02": "host02" }
```

You could use the `get` function to dynamically pull the value from your `hosts` object:

```coffee
$ get!(hosts, [.serialnumber])
"host01"
```

**Reply by StefanSa** [reply]
@spencergilbert 
hi Spencer,
as always thanks for your quick help.
The solution looks simple, but I can't get the object `.host.name` created. Apparently the SN is not found, maybe a path problem ? 
```
hosts = { "C241B":"fw01", "C24DA":"fw02" }

.host.name =  get!(hosts, [.sophos.xg.device_serial_id])
```
i am currently doing this.
```
if contains(.sophos.xg.device_serial_id, "C24V1B") {
  .host.name = "fw01"
}
```

**Reply by spencergilbert** [reply]
I'd suggest using the [`log` ](https://vector.dev/docs/reference/vrl/functions/#log) function to print the event, or parts of the event that you are working on. If that path exists and matches a key in the `hosts` object it should work.

**Reply by StefanSa** [reply]
@spencergilbert Spencer,
In the stdout the host is correctly read from the variable `Internal log ["fw01"]`. However, it is not transferred to `.host.name`.

```
.host.name =  get!(hosts, [.sophos.xg.device_serial_id])
log(.host.name)
```

**Reply by spencergilbert** [reply]
I'm not sure where this is coming from, or what is producing this. Can you explain more?

>In the stdout the host is correctly read from the variable Internal log ["fw01"].

**Reply by StefanSa** [reply]
@spencergilbert Spencer,
I can't really explain this, or what do you want to know to understand this problem ?
We have this variable:
`hosts = { "C241B":"fw01", "C24DA":"fw02" }`

And read the hostname with `get`.
```
.host.name =  get!(hosts, [.sophos.xg.device_serial_id])
```
according to `log debug` and clearly visible in the `stdout` the host is read correctly from the variable.
`log(.host.name)` => `Internal log ["fw01"]`

The question is, why does the function `get` transfer the correctly identified value not to `.host.name`?
Spencer, unfortunately, that's all i can say about it.

**Reply by spencergilbert** [reply]
If the following prints the correct hostname to Vector's logs then it is working and is setting the `.host.name` path based on the `get` function:

```coffee
hosts = { "C241B":"fw01", "C24DA":"fw02" }
.host.name =  get!(hosts, [.sophos.xg.device_serial_id])
log(.host.name)
```

If `.host.name` wasn't being set you'd have something that looked like:

```
2022-01-13T14:57:03.609761Z  INFO vrl_stdlib::log: null internal_log_rate_secs=1 vrl_position=0
2022-01-13T14:57:03.610275Z  INFO vrl_stdlib::log: Internal log [null] is being rate limited.
```

I suspect you'll be able to better get to the bottom of your problems if you swap the `!` abort argument from your functions and add more robust handling like:

```coffee
hosts = { "C241B":"fw01", "C24DA":"fw02" }
.host.name, err =  get(hosts, [.sophos.xg.device_serial_id])
if err != null {
  log("Hostname not found for serial id: " + { string(.sophos.xg.device_serial_id) ?? "`.sophos.xg.device_serial_id` is not a string" }, level: "error")
  abort
}
```

**Reply by StefanSa** [reply]
@spencergilbert 
Spencer nope, no error is returned.
This is a hard nut to crack.

**Reply by spencergilbert** [reply]
If no error is logged than the error didn't occur and the field was properly set. Please share the entirety of your config here and point out where you think the field isn't available.

**Reply by StefanSa** [reply]
@spencergilbert Spencer
you got a PM (gmail.com) from me.

**Reply by spencergilbert** [reply]
Received! (I'm also available under the same email but @datadog.com). I'll take a look and see what I can do this afternoon.

**Reply by spencergilbert** [reply]
@StefanSa it looks like the config you shared is running `del(.host)` in a couple of places _after_ setting the field via `get()`.

```bash
❯ grep 'del(.host)' *.toml
***.toml:.url.domain = del(.host)
***.toml:.url.domain = del(.host)
***.toml:.url.domain = del(.host)
***.toml:del(.host)
```

**Reply by StefanSa** [reply]
> @StefanSa it looks like the config you shared is running `del(.host)` in a couple of places _after_ setting the field via `get()`.
> 
> 
> 
> ```bash
> 
> ❯ grep 'del(.host)' *.toml
> 
> ***.toml:.url.domain = del(.host)
> 
> ***.toml:.url.domain = del(.host)
> 
> ***.toml:.url.domain = del(.host)
> 
> ***.toml:del(.host)
> 
> ```

Hmm, then the question is why does it work with both "If" queries currently commented out for testing.

**Reply by spencergilbert** [reply]
It looks like the `.host` key will be removed before the final output regardless. Can you put together a minimal example config and event that reproduces the error you're seeing that I can run? It's difficult to reproduce the issue with your large config and no inputs to test with.

**Reply by StefanSa** [reply]
@spencergilbert 
Oh man, you are right about `del(.host)`, i was completely blind here. :(
Everything works as it should. Sorry for your time.
But one question remains.
`parse_url` itself publishes a `.host` and overwrites me the key, how can i prevent that ?

**Reply by jszwedko** [reply]
Assuming you are merging the result of `parse_url` into your event, I'd recommend deleting or renaming it. You could consider something like:

```
url_parts = parse_url!(.some_url)
url_parts.url_host = del(url_parts.host)
. |= url_parts
```

**Reply by StefanSa** [reply]
Hi Jesse and Spencer you are a great team, thanks for the support here.
I'm off work soon and can't get this together, what I finally have to do here. Please help again, thank you !
```
if exists(.sophos.xg.url) {
   .url.original = del(.sophos.xg.url)
. |= parse_url!(.url.original, default_known_ports: true)
.url.domain = .host
.url.domain = del(.host)
.url.path = del(.path)
.url.scheme = del(.scheme)
.url.fragment = del(.fragment)
.url.password = del(.password)
.url.username = del(.username)
.url.port = del(.port)
.url.query = del(.query)
}
```
Have a nice weekend in advance.

**Reply by jszwedko** [reply]
Hi @StefanSa . As mentioned above, merging the result of `parse_url` directly into `.` will cause it to override `host`. You'll want to save the result of `parse_url` to an intermediate variable and then assign from there. For example:

```
url_parts = parse_url!(.some_url)
.url.domain = url_parts.host
```

**Reply by StefanSa** [reply]
Jesse perfect. Thanks
